### PR TITLE
feat: support multiple resource definition of the same type

### PR DESCRIPTION
**Pull request by gitlawr**
Related issue: https://github.com/seal-io/walrus/issues/1784

### CODE REVIEW

File: pkg/apis/environment/common.go [hunk 1]
```diff
@@ -93,6 +93,12 @@ func createEnvironment(
 }
 
 func validateEnvironmentCreateInput(r model.EnvironmentCreateInput) error {
+	if err := r.Validate(); err != nil {
+		return err
+	}
+
+	cache := make(map[string]*model.ResourceDefinition)
```
Comment (aiwantaozi) [inline pkg/apis/environment/common.go:100]: No one has been found writing to this cache.

File: pkg/apis/resource/basic_view.go [hunk 2]
```diff
@@ -219,19 +218,25 @@ type (
...
-				Type: item.Type,
-			}
-		}
-	}
-
 	if err := r.ResourceCreateInputs.Validate(); err != nil {
 		return err
 	}
 
+	cache := make(map[string]*model.ResourceDefinition)
```
Comment (aiwantaozi) [inline pkg/apis/resource/basic_view.go:225]: ditto

File: pkg/resourcedefinitions/match.go [hunk 3]
```diff
@@ -64,3 +71,20 @@ func matchLabels(selectors, labels map[string]string) bool {
...
+		return errors.New("both are applicable to all projects")
+	}
+
+	for _, pName1 := range pNames1 {
+		for _, pName2 := range pNames2 {
+			if pName1 == pName2 {
+				return fmt.Errorf("both are applicable to project %q", pName1)
+			}
+		}
+	}
```
Comment (thxCode) [inline pkg/resourcedefinitions/match.go:87]: speed up by sets.

File: pkg/apis/environment/extension.go [hunk 4]
```diff
@@ -238,7 +240,37 @@ func (h Handler) RouteClone(req RouteCloneEnvironmentRequest) (*RouteCloneEnviro
...
+				OrderExpr(
+					sql.Expr(
+						fmt.Sprintf(
+							`%s, jsonb_array_length(%s) DESC`,
+							resourcedefinition.FieldType,
+							resourcedefinition.FieldApplicableProjectNames,
+						),
+					),
+				)
+		}).
```
Comment (thxCode) [inline pkg/apis/environment/extension.go:273]: please give a comment to describe the final result.

File: pkg/apis/environment/common.go (hunk 1 above)
Comment (thxCode) [reply pkg/apis/environment/common.go:100]: @aiwantaozi  it seems that the `dao.GetMatchingResourceDefinitionWith` needs a `cache`.

File: pkg/dao/resource_definition.go [hunk 5]
```diff
@@ -85,3 +88,67 @@ func ResourceDefinitionMatchingRulesEdgeSave(
...
+				OrderExpr(
+					sql.Expr(
+						fmt.Sprintf(
+							`%s, jsonb_array_length(%s) DESC`,
+							resourcedefinition.FieldType,
+							resourcedefinition.FieldApplicableProjectNames,
+						),
+					),
+				)
+		}).
```
Comment (thxCode) [inline pkg/dao/resource_definition.go:147]: please give a comment to describe the final result.

File: pkg/dao/resource_definition.go [hunk 6]
```diff
@@ -6,6 +6,7 @@ import (
 	"errors"
 
 	"entgo.io/ent/dialect/sql"
+
```
Comment (thxCode) [inline pkg/dao/resource_definition.go:9]: remove this line

**Review by thxCode — DISMISSED** [review]
~LGTM~

**Review by shanewxy — DISMISSED** [review]
LGTM

**Comment by gitlawr** [comment]
Rebased

**Review by thxCode — APPROVED** [review]
LGTM

**Review by aiwantaozi — APPROVED** [review]
LGTM

**Pull request merged**